crates/build/src/solidity_metadata/abi.rs

```diff
@@ -0,0 +1,377 @@
+// Copyright (C) ink! contributors.
+// This file is part of cargo-contract.
+//
+// cargo-contract is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// cargo-contract is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with cargo-contract.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::{
+    collections::BTreeMap,
+    fs::{
+        self,
+    },
+    path::{
+        Path,
+        PathBuf,
+    },
+};
+
+use alloy_json_abi::{
+    Constructor,
+    Event,
+    Function,
+    JsonAbi,
+};
+use anyhow::Result;
+use ink_metadata::{
+    ConstructorSpec,
+    EventSpec,
+    InkProject,
+    MessageParamSpec,
+    MessageSpec,
+    ReturnTypeSpec,
+};
+use itertools::Itertools;
+use scale_info::{
+    form::PortableForm,
+    PortableRegistry,
+    TypeDef,
+    TypeDefPrimitive,
+};
+
+use crate::CrateMetadata;
+
+/// Generates a Solidity-compatible ABI for the ink! smart contract (if possible).
+///
+/// Ref: <https://docs.soliditylang.org/en/latest/abi-spec.html#abi-json>
+pub fn generate_abi(ink_project: &InkProject) -> Result<JsonAbi> {
+    let registry = ink_project.registry();
+    let spec = ink_project.spec();
+
+    // Solidity allows only one constructor, so we arbitrarily choose the first one.
+    // FIXME: (@davidsemakula) can we make constructor selection a bit more deterministic?
```

Comment:
ink! supports the annotation of a default constructor with the syntax

```rust
#[ink(constructor, default)]
pub fn my_constructor() -> Self {}
```

[Code pointer](https://github.com/use-ink/ink/blob/7338425630ca122717e360573d1b6ad09992d559/crates/ink/ir/src/ir/attrs.rs#L301-L305)

Can you take that into consideration here? So that if a default one exists we use that one. 

The behavior of how the constructor is chosen should also be documented in the comments.

Reply:
Makes sense, didn't think about the default arg for some reason 🙂

Reply:
Done